In [2]:
import os
os.chdir('..') if os.getcwd().endswith('notebooks') else None

import pandas as pd
from sqlalchemy import create_engine
from dotenv import load_dotenv

load_dotenv()
DB_PASSWORD = os.getenv("DB_PASSWORD")
engine = create_engine(f"postgresql+psycopg2://postgres:{DB_PASSWORD}@localhost:5432/RetailPulse")

query = """
SELECT o.invoice, o.country, SUM(oi.quantity * oi.price) AS order_value
FROM order_items oi
JOIN orders o USING(invoice)
GROUP BY o.invoice, o.country
"""
orders = pd.read_sql(query, engine)
orders['is_uk'] = orders['country'] == 'United Kingdom'

orders.groupby('is_uk')['order_value'].describe()

,count,mean,std,min,25%,50%,75%,max
is_uk,,,,,,,,
False,2182.0,615.979743,1772.257998,-11880.84,15.0,320.32,732.8575,44051.60
True,23260.0,359.789456,982.999334,-25111.09,45.0,219.68,401.6625,49844.99


In [3]:
from scipy import stats

uk_values = orders[orders['is_uk'] == True]['order_value']
non_uk_values = orders[orders['is_uk'] == False]['order_value']

t_stat, p_value = stats.ttest_ind(uk_values, non_uk_values, equal_var=False)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.10f}")

t-statistic: -6.6571
p-value: 0.0000000000


In [4]:
print(f"p-value: {p_value:.2e}")

p-value: 3.48e-11


In [5]:
mean_diff = uk_values.mean() - non_uk_values.mean()
print(f"Mean difference (UK - Non-UK): ₹{mean_diff:.2f}")

# 95% confidence interval for the difference in means
ci = stats.ttest_ind(uk_values, non_uk_values, equal_var=False).confidence_interval(confidence_level=0.95)
print(f"95% CI for the difference: (₹{ci.low:.2f}, ₹{ci.high:.2f})")

Mean difference (UK - Non-UK): ₹-256.19
95% CI for the difference: (₹-331.66, ₹-180.72)


Finding: UK and non-UK customers show a statistically significant difference in average order value (Welch's t-test, t=-6.66, p=3.48×10⁻¹¹). Non-UK orders average ₹256 more than UK orders (95% CI: ₹181–₹332). This difference is highly unlikely to be due to random chance.

Caveat: This test establishes that a difference exists and its approximate size — it does not explain why. The cause may be order frequency, customer mix (retail vs. wholesale), or other factors not yet investigated.


In [6]:
result = stats.ttest_ind(uk_values, non_uk_values, equal_var=False)
print(f"Degrees of freedom: {result.df:.2f}")

Degrees of freedom: 2308.52


CHI- SQUARE TEST to determine whether country is associated with number of Cancelled order.

In [7]:
from scipy.stats import chi2_contingency
import numpy as np

contingency_table = np.array([
    [4005, 19255],   # UK: cancelled, not_cancelled
    [522, 1660]       # non-UK: cancelled, not_cancelled
])

chi2_stat, p_value, dof, expected = chi2_contingency(contingency_table)

print(f"Chi-square statistic: {chi2_stat:.4f}")
print(f"p-value: {p_value:.10f}")
print(f"Degrees of freedom: {dof}")
print(f"Expected counts (if H0 were true):\n{expected}")

Chi-square statistic: 60.8472
p-value: 0.0000000000
Degrees of freedom: 1
Expected counts (if H0 were true):
[[ 4138.74773996 19121.25226004]
 [  388.25226004  1793.74773996]]


In [8]:
n = contingency_table.sum()
cramers_v = np.sqrt(chi2_stat / (n * (min(contingency_table.shape) - 1)))
print(f"Cramér's V: {cramers_v:.4f}")

Cramér's V: 0.0489


In [9]:
Q1 = orders['order_value'].quantile(0.25)
Q3 = orders['order_value'].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"Q1: {Q1:.2f}, Q3: {Q3:.2f}, IQR: {IQR:.2f}")
print(f"Lower bound: {lower_bound:.2f}, Upper bound: {upper_bound:.2f}")

anomalies = orders[(orders['order_value'] < lower_bound) | (orders['order_value'] > upper_bound)]
print(f"Number of flagged anomalies: {len(anomalies)} out of {len(orders)} ({len(anomalies)/len(orders)*100:.2f}%)")

Q1: 42.00, Q3: 419.52, IQR: 377.52
Lower bound: -524.28, Upper bound: 985.79
Number of flagged anomalies: 2039 out of 25442 (8.01%)


In [10]:
low_anomalies = orders[orders['order_value'] < lower_bound]
high_anomalies = orders[orders['order_value'] > upper_bound]

print(f"Low-end anomalies: {len(low_anomalies)}")
print(f"High-end anomalies: {len(high_anomalies)}")

high_anomalies.sort_values('order_value', ascending=False).head(10)

Low-end anomalies: 166
High-end anomalies: 1873


,invoice,country,order_value,is_uk
24596,533027,United Kingdom,49844.99,True
24495,531516,United Kingdom,45332.97,True
17527,493819,EIRE,44051.60,False
13722,524181,United Kingdom,33167.80,True
12144,526934,United Kingdom,26007.08,True
11762,512771,United Kingdom,25111.09,True
1057,515944,United Kingdom,22863.36,True
6969,503061,United Kingdom,22853.77,True
4208,517731,United Kingdom,21984.00,True
4996,495591,Netherlands,18570.88,False


In [11]:
query = """
SELECT o.invoice, o.country, p.description, oi.quantity, oi.price, (oi.quantity * oi.price) AS line_value
FROM order_items oi
JOIN orders o USING(invoice)
JOIN products p ON oi.stock_code = p.stock_code
WHERE o.invoice = '533027'
ORDER BY line_value DESC;
"""
pd.read_sql(query, engine)

,invoice,country,description,quantity,price,line_value
0,533027,United Kingdom,PAPER CHAIN KIT 50'S CHRISTMAS,835,6.95,5803.25
1,533027,United Kingdom,ROTATING SILVER ANGELS T-LIGHT HLDR,441,4.95,2182.95
2,533027,United Kingdom,PAPER CHAIN KIT VINTAGE CHRISTMAS,290,6.95,2015.50
3,533027,United Kingdom,"POPCORN HOLDER , SMALL",902,1.95,1758.90
4,533027,United Kingdom,3 HEARTS HANGING DECORATION RUSTIC,242,6.95,1681.90
...,...,...,...,...,...,...
106,533027,United Kingdom,CARD CHRISTMAS VILLAGE,40,1.00,40.00
107,533027,United Kingdom,LIPSTICK PEN FUSCHIA,36,1.00,36.00
108,533027,United Kingdom,SWALLOW WOODEN CHRISTMAS DECORATION,18,1.95,35.10
109,533027,United Kingdom,CIRCUS PARADE LUNCHBOX,7,4.95,34.65


In [12]:
orders['is_anomaly'] = (orders['order_value'] < lower_bound) | (orders['order_value'] > upper_bound)

anomaly_summary = orders.groupby('is_anomaly')['order_value'].agg(['count', 'mean', 'sum'])
print(anomaly_summary)

            count         mean         sum
is_anomaly                                
False       23403   236.693732  5539343.40
True         2039  2046.800956  4173427.15


IQR-based anomaly detection flags 2,039 orders (8.01%) as statistical outliers by order value. Manual investigation of the top flagged order (533027, ₹49,845, 107 line items) confirms these are predominantly legitimate bulk/wholesale purchases, not fraud or data errors. This flagged segment represents a disproportionate 42.97% of total revenue despite being a small fraction of order count — suggesting a distinct high-value customer segment worth identifying and serving deliberately (e.g., dedicated wholesale pricing or account management), rather than being treated as noise. Further work: cross-reference flagged orders with customer IDs to see if this is a small, identifiable set of repeat wholesale buyers or distributed across many one-time bulk purchases.

In [19]:
from sqlalchemy import text

query = """
SELECT 
    o.customer_id,
    o.country,
    COUNT(DISTINCT o.invoice) AS frequency,
    SUM(oi.quantity * oi.price) AS total_revenue,
    AVG(oi.quantity * oi.price) AS avg_order_value,
    MAX(o.invoice_date) AS last_purchase_date,
    MIN(o.invoice_date) AS first_purchase_date,
    SUM(CASE WHEN o.invoice LIKE 'C%' THEN 1 ELSE 0 END) AS cancellations,
    AVG(oi.quantity) AS avg_quantity_per_item
FROM orders o
JOIN order_items oi USING(invoice)
WHERE o.customer_id IS NOT NULL
GROUP BY o.customer_id, o.country
"""

customers = pd.read_sql(text(query), engine)

# Convert dates before doing date arithmetic
customers['last_purchase_date'] = pd.to_datetime(
    customers['last_purchase_date']
)

customers['first_purchase_date'] = pd.to_datetime(
    customers['first_purchase_date']
)

# Recency
dataset_end = customers['last_purchase_date'].max()

customers['recency_days'] = (
    dataset_end - customers['last_purchase_date']
).dt.days

# Customer tenure
customers['tenure_days'] = (
    customers['last_purchase_date'] -
    customers['first_purchase_date']
).dt.days

customers.describe()

,customer_id,frequency,total_revenue,avg_order_value,last_purchase_date,first_purchase_date,cancellations,avg_quantity_per_item,recency_days,tenure_days
count,4384.000000,4384.000000,4384.000000,4384.000000,4384,4384,4384.000000,4384.000000,4384.000000,4384.000000
mean,15347.191834,5.370666,1896.549286,13.352337,2010-09-09 13:51:21,2010-04-25 12:15:06,2.236770,19.680564,90.422673,137.066834
min,12346.000000,1.000000,-25111.090000,-25111.090000,2009-12-01 00:00:00,2009-12-01 00:00:00,0.000000,-16.000000,0.000000,0.000000
25%,13874.750000,1.000000,283.595000,10.043340,2010-07-25 00:00:00,2010-01-09 12:00:00,0.000000,4.886012,16.000000,0.000000
50%,15351.500000,3.000000,652.575000,16.393585,2010-10-19 00:00:00,2010-03-29 00:00:00,0.000000,8.791667,51.000000,110.000000
75%,16836.500000,6.000000,1634.727500,22.905366,2010-11-23 00:00:00,2010-08-11 00:00:00,2.000000,13.595231,137.000000,260.000000
max,18287.000000,269.000000,341776.730000,2803.200000,2010-12-09 00:00:00,2010-12-09 00:00:00,171.000000,3990.684211,373.000000,373.000000
std,1702.967643,10.023230,8515.674386,421.255210,NaN,NaN,7.217431,106.143276,98.129978,134.053757
